docs/reference/rego-templates.md

```diff
@@ -0,0 +1,214 @@
+
+```rego
+package ratify.policy
+default valid := false
+valid if {
+	not failed_verify(input)
+}
+failed_verify(reports) if {
+	newReports := {"nestedReports": reports.verifierReports}
+	has_subject_failed_verify(newReports)
+}
+has_subject_failed_verify(nestedReports) if {
+	[path, value] := walk(nestedReports)
+	not artifact_type_pass_verify(value)
+	path[count(path) - 1] == "nestedReports"
+}
+# at least one artifact of the same type passed verification
+artifact_type_pass_verify(nestedReports) if {
+	count_artifact_type(nestedReports) == count_successful_artifact_type(nestedReports)
+}
+count_artifact_type(nestedReports) := number if {
+	artifact_types := {x |
+		some i
+		x := nestedReports[i].artifactType
+	}
+	number := count(artifact_types)
+}
+count_successful_artifact_type(nestedReports) := number if {
+	artifact_types := {x |
+		some i
+		x := nestedReports[i].artifactType
+		artifact_pass_verify(nestedReports[i].verifierReports)
+	}
+	number := count(artifact_types)
+}
+# an artifact has at least one successful report
+artifact_pass_verify(verifierReports) if {
+	verifierReports[_].isSuccess == true
+}
+# should be at least one verifier report that is successful
+failed_verify(reports) if {
+	[path, value] := walk(reports)
+	path[count(path) - 1] == "verifierReports"
+	count(value) == 0
+}
+```
+
+### Example 3
+I trust multiple identities, and need to find at least one valid signature signed by any identity.
+
+Prerequisites:
+
+1. The artifact is only signed by Notation, there might be a few signatures attached to the artifact.
+2. Users has configured different Notation verfiers, and each verifier has a different identity.
+
+
+```rego
+package ratify.policy
+import future.keywords.in
+default valid := false
+valid {
+  not failed_verify(input)
+}
+# Fail the verification if no signatures are signed by trusted identity.
+failed_verify(reports) {
+  not has_trusted_signature(reports.verifierReports)
+}
+# Fail the verification if no signatures are present.
+failed_verify(reports) {
+  [path, value] := walk(reports)
+  path[count(path) - 1] == "verifierReports"
+  count(value) == 0
+}
+# There is a signature signed by trusted identity.
+has_trusted_signature(reports) {
+	some report in reports
+  report.artifactType == "application/vnd.cncf.notary.signature"
+  pass_validation(report.verifierReports)
+}
+pass_validation(reports) {
+	some report in reports
+  report.isSuccess == true
+}
+```
+
+### Example 4
+I trust multiple identities, and need to find at least one valid signature signed by each identity.
+
+Prerequisites:
+
+1. The artifact is only signed by Notation, there might be a few signatures attached to the artifact.
+2. Users has configured different Notation verfiers, and each verifier has a different identity.
+
+Note: The below Rego query currently does not work as expected as it's blocked by a known [bug](https://github.com/deislabs/ratify/issues/1066) which reports the same name for different Notation verifiers.
```

Comment:
Do you think we should wait to add this example until after 1066 has been addressed?

Reply:
I agree with you waiting until 1066 gets fixed. @FeynmanZhou, as we discussed in previous meetings, we'd like to have these examples in v1. Do you think we can wait for next release?

Reply:
btw, I have removed this example, I'll add it back once we fix the blocking issue.